controllers/linstorcluster_controller.go

```diff
@@ -0,0 +1,571 @@
+	imgs, err := r.ImageVersions.GetVersions(lcluster.Spec.Repository, "")
+	if err != nil {
+		return nil, err
+	}
+
+	saPatch, err := r.pullSecretPatch()
+	if err != nil {
+		return nil, err
+	}
+
+	k := &kusttypes.Kustomization{
+		Namespace: r.Namespace,
+		Labels:    r.kustomLabels(lcluster),
+		Resources: []string{"controller"},
+		Images:    imgs,
+		Patches:   append(utils.MakeKustPatches(lcluster.Spec.Patches...), saPatch...),
+	}
+
+	return r.Kustomizer.Kustomize(k)
+}
+
+// Create the CSI controller and node agent resources.
+//
+// Applies the following changes over the base resources:
+// * Namespace
+// * default labels
+// * default images
+// * pull secret (if any)
+// * restrict CSI driver daemon set to cluster's node selector
+// * user defined patches
+func (r *LinstorClusterReconciler) kustomizeCsiResources(lcluster *piraeusiov1.LinstorCluster) (resmap.ResMap, error) {
+	selectorPatch, err := utils.ToEncodedPatch(
+		&kusttypes.Selector{ResId: resid.NewResId(resid.NewGvk(appsv1.GroupName, "v1", "DaemonSet"), "csi-node")},
+		applyappsv1.DaemonSet("satellite", "").
+			WithSpec(applyappsv1.DaemonSetSpec().
+				WithTemplate(applycorev1.PodTemplateSpec().
+					WithSpec(applycorev1.PodSpec().WithNodeSelector(lcluster.Spec.NodeSelector))),
+			),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	imgs, err := r.ImageVersions.GetVersions(lcluster.Spec.Repository, "")
+	if err != nil {
+		return nil, err
+	}
+
+	saPatch, err := r.pullSecretPatch()
+	if err != nil {
+		return nil, err
+	}
+
+	k := &kusttypes.Kustomization{
+		Namespace: r.Namespace,
+		Labels:    r.kustomLabels(lcluster),
+		Resources: []string{"csi"},
+		Images:    imgs,
+		Patches:   append(append(utils.MakeKustPatches(lcluster.Spec.Patches...), saPatch...), *selectorPatch),
+	}
+
+	return r.Kustomizer.Kustomize(k)
+}
+
+// Create the common resources for LINSTOR satellites, but not the actual LinstorSatellite resources.
+//
+// The resources here are shared by all LinstorSatellite instances. This is used for:
+// * A common ServiceAccount, with optional pull secret configured
+// * A NetworkPolicy to protect DRBD ports from unauthorized access.
+//
+// Applies the following changes over the base resources:
+// * Namespace
+// * default labels
+// * default images
+// * pull secret (if any)
+// * user defined patches
+func (r *LinstorClusterReconciler) kustomizeNodeCommonResources(lcluster *piraeusiov1.LinstorCluster) (resmap.ResMap, error) {
+	imgs, err := r.ImageVersions.GetVersions(lcluster.Spec.Repository, "")
+	if err != nil {
+		return nil, err
+	}
+
+	saPatch, err := r.pullSecretPatch()
+	if err != nil {
+		return nil, err
+	}
+
+	k := &kusttypes.Kustomization{
+		Namespace: r.Namespace,
+		Labels:    r.kustomLabels(lcluster),
+		Resources: []string{"satellite-common"},
+		Images:    imgs,
+		Patches:   append(utils.MakeKustPatches(lcluster.Spec.Patches...), saPatch...),
+	}
+
+	return r.Kustomizer.Kustomize(k)
```

Comment:
These lines are basically repeated 3 times. Extract into separate function? Or do `GetVersions` and `pullSecretPatch` once and pass the values in?

Reply:
Done. This led down an interesting rabbit hole, because kustomize wanted to set the namespace on `LinstorSatellite` resources, which is fine for applying, but tripped up our resource pruning code, as it thought "namespaced satellite != cluster-scoped satellite" and immediately tried to remove the satellite after applying